x/protorev/keeper/rebalance.go

```diff
@@ -94,28 +104,50 @@ func (k Keeper) EstimateMultihopProfit(ctx sdk.Context, inputDenom string, amoun
 }
 
 // FindMaxProfitRoute runs a binary search to find the max profit for a given route
-func (k Keeper) FindMaxProfitForRoute(ctx sdk.Context, route poolmanagertypes.SwapAmountInRoutes, inputDenom string) (sdk.Coin, sdk.Int, error) {
+func (k Keeper) FindMaxProfitForRoute(ctx sdk.Context, route RouteMetaData, remainingPoolPoints *uint64) (sdk.Coin, sdk.Int, error) {
+	// Track the tokenIn amount/denom and the profit
 	tokenIn := sdk.Coin{}
 	profit := sdk.ZeroInt()
 
+	// Track the left and right bounds of the binary search
 	curLeft := sdk.OneInt()
 	curRight := types.MaxInputAmount
-	iteration := 0
 
-	for curLeft.LT(curRight) && iteration < types.MaxIterations {
-		iteration++
+	// Input denom used for cyclic arbitrage
+	inputDenom := route.Route[route.Route.Length()-1].TokenOutDenom
+
+	// If a cyclic arb exists with an optimal amount in above our minimum amount in,
+	// then inputting the minimum amount in will result in a profit. So we check for that first.
+	// If there is no profit, then we can return early and not run the binary search.
+	_, minInProfit, err := k.EstimateMultihopProfit(ctx, inputDenom, curLeft.Mul(route.StepSize), route.Route)
+	if err != nil {
+		return sdk.Coin{}, sdk.ZeroInt(), err
+	} else if minInProfit.LTE(sdk.ZeroInt()) {
+		return sdk.Coin{}, sdk.ZeroInt(), fmt.Errorf("no profit for route %d", route.Route.PoolIds())
```

Comment:
I imagine a majority of small swaps will result in no profits, so this will throw an error often. Feels kinda like we want to view no profit in a route as a non-error. maybe we log in some other way if we want to log?

Reply:
Resolved in ad5ea732f09ea71e7433825dba9f6523223c524d